File: integration-test/src/test/js/cnft-common-steps.js
const { fail } = require('assert');
const { CNFT, CreateProfileInfo, TokenContent, WalletIssueTokenRequest } = require('wallet-lib');
const { Ascii, createWalletCrypto, log, registerMemberHook, registerTokenType } = require('./cnft-util');

const { FgGreen: FgBase, FgMagenta: FgVal, FgYellow: FgEvt, Reset } = Ascii;

const walletContextMixin = {

    wallets: {},
    walletIdentities: {},

    getWalletByClientName(clientName) {
        const wallet = this.wallets[clientName];
        if (typeof wallet === 'undefined') {
            fail(`Unknown client ${clientName}`);
        }
        return wallet;
    },

    getWalletIdentityByClientName(clientName) {
        const walletIdentity = this.walletIdentities[clientName];
        if (typeof walletIdentity === 'undefined') {
            fail(`Unknown client ${clientName}`);
        }
        return walletIdentity;
    },

};

const tokenTypesContextMixin = {

    tokenTypeIds: {},

    getTokenTypeIdByName(tokenTypeName) {
        const typeId = this.tokenTypeIds[tokenTypeName];
        if (typeof typeId === 'undefined') {
            fail(`Unknown token type ${tokenTypeName}`);
        }
        return typeId;
    },

};

const tokensContextMixin = {

    tokenIds: {},

    getTokenIdByName(tokenName) {
        const token = this.tokenIds[tokenName];
        if (typeof token === 'undefined') {
            fail(`Unknown token name ${tokenName}`);
        }
        return token;
    },

};

const profilesContextMixin = {

    profiles: {},

    getProfileByName(profileName) {
        const profile = this.profiles[profileName];
        if (typeof profile === 'undefined') {
            fail(`Unknown profile name ${profileName}`);
        }
        return profile;
    },

};

const CNFTGateUrl = 'http://localhost:8981';
const Chain = CNFT.connect(CNFTGateUrl);

async function awaitWalletEvents(wallet, f) {
    var running = true;
    var lastSeenBlock = await wallet.chain.getLatestBlockNumber;
    while (running) {
        let height = await wallet.chain.getLatestBlockNumber;
        while (lastSeenBlock <= height) {
            let events = await wallet.events(lastSeenBlock, false);
            let result = f(wallet, events);
            if(result)
                return events;
            lastSeenBlock = (parseInt(lastSeenBlock) + 1);
        }
        await new Promise((resolve) => {
            setTimeout(resolve, 1000);
        });
    }
}

function waitWalletEvent(wallet, eventType) {

    awaitWalletEvents(wallet, (wallet, events) => {
        let eventResult;
        switch (eventType) {
            case 'tokensReceived':
                eventResult = events.owner.tokensReceived;
                break;
            case 'tokensBurn':
                eventResult = events.owner.tokensBurn;
                break;
            case 'created':
                eventResult = events.profiles.created;
                break;
            case 'updated':
                eventResult = events.profiles.updated;
                break;
            case 'tokensLinked':
                eventResult = events.profiles.tokensLinked;
                break;
            case 'tokensUnlinked':
                eventResult = events.profiles.tokensUnlinked;
                break;
            default:
                throw(`Error: "${eventType}" is not correct eventType parameter for waitProfileEvent() function`)
        }

        if (eventResult.length > 0) {
            log(`${FgVal}${eventType}=${Reset}[${eventResult}]\n`);
            return true;
        }
        return false;
    });
}

async function thereIsAClient(context, clientName) {
    log(`${FgBase}There is a client ${FgVal}'${clientName}'${Reset}`);

    log(`${FgBase}Creating a wallet configuration for client ${FgVal}'${clientName}'${FgBase}...${Reset}`);
    log(`${FgBase}  Done: created a wallet configuration for client ${FgVal}'${clientName}'${Reset}`);

    log(`${FgBase}Creating a new wallet for client ${FgVal}'${clientName}'${FgBase}...${Reset}`);

    const crypto = await createWalletCrypto();
    const wallet = await Chain.newWallet(crypto);
    await registerMemberHook(crypto);

    context.wallets[clientName] = wallet;
    context.walletIdentities[clientName] = await wallet.getIdentity;

    log(`${FgBase}  Done: created a new wallet for client ${FgVal}'${clientName}'${Reset}\n`);
}

async function clientRegistersTokenType(context, clientName, tokenTypeName, gene) {
    log(`${FgBase}Client ${FgVal}'${clientName}'${FgBase} registers token type ${FgVal}'${tokenTypeName}'${Reset}`);
    context.tokenTypeIds[tokenTypeName] = await registerTokenType(
        context.getWalletByClientName(clientName),
        tokenTypeName,
        typeof gene !== 'undefined' ? [ gene ] : []
    );
    log(`${FgBase}  Done: typeId=${FgVal}'${context.getTokenTypeIdByName(tokenTypeName)}'${Reset}\n`);
}

async function clientCreatesIdForTokenOfType(context, clientName, tokenName, tokenTypeName) {
    log(`${FgBase}Client ${FgVal}'${clientName}'${FgBase} creates ID for token ${FgVal}'${tokenName}'` +
        `${FgBase} of type ${FgVal}'${tokenTypeName}'${Reset}`);

    const wallet = context.getWalletByClientName(clientName);
    const tokenTypeId = context.getTokenTypeIdByName(tokenTypeName);
    try {
        const tokenId = await wallet.createTokenId(tokenTypeId);
        context.tokenIds[tokenName] = tokenId;
        log(`${FgBase}  Done: tokenId=${FgVal}'${tokenId}'${Reset}\n`);
    } catch (error) {
        log(error);
        throw error;
    }
}

async function clientIssuesTokenForClientWithValue(context, issuerName, clientName, tokenName, tokenValue) {
    const issuerWallet = context.getWalletByClientName(issuerName);
    const clientWallet = context.getWalletByClientName(clientName);
    const tokenId = context.getTokenIdByName(tokenName);

    const owner = await clientWallet.createSingleOwnerAddress;
    const to = await clientWallet.getIdentity;
    const body = TokenContent([ tokenValue ]);
    const request = WalletIssueTokenRequest(tokenId, owner, body, to);
    await issuerWallet.issue([ request ]).catch(function (error){
        log(error);
        throw (error);
    });
}

async function clientCreatesProfile(
    context,
    clientName,
    createProfileInfo = CreateProfileInfo(
        'test_name',
        'test_description',
        'test_avatar',
        'test_background'
    )
) {
    log(`${FgBase}Client ${FgVal}'${clientName}'${FgBase} creates profile ${FgVal}'${createProfileInfo.name}'${Reset}`);
    const wallet = await context.getWalletByClientName(clientName);

    const waitProfilesCreated = awaitWalletEvents(wallet, (wallet, events) => {
        if (events.profiles.created.length > 0) {
            log(`Profiles Created=[${events.profiles.created}]`);
            return true;
        }
        return false;
    })

    const tx = await wallet.createProfile(createProfileInfo);

    context.profiles[createProfileInfo.name] = tx.value;
    log(`${FgBase}  Done: profile.id=${FgVal}'${tx.value.id}'${Reset}\n`);

    return waitProfilesCreated;
}

module.exports = {
    clientCreatesIdForTokenOfType,
    clientCreatesProfile,
    clientIssuesTokenForClientWithValue,
    clientRegistersTokenType,
    thereIsAClient,
    awaitWalletEvents,
    waitWalletEvent,
    profilesContextMixin,
    tokenTypesContextMixin,
    tokensContextMixin,
    walletContextMixin,
};

File: integration-test/src/test/scala/ru/sberbank/blockchain/cnft/wallet/test/CNFTCommonSteps.scala
package ru.sberbank.blockchain.cnft.wallet.test

import ru.sberbank.blockchain.cnft.common.types.{Collection, Result}
import ru.sberbank.blockchain.cnft.model._
import ru.sberbank.blockchain.cnft.wallet.dsl._
import ru.sberbank.blockchain.cnft.wallet.spec.CNFTWalletSpec
import ru.sberbank.blockchain.cnft.wallet.test.CNFTUtil._
import ru.sberbank.blockchain.cnft.wallet.walletmodel.WalletIssueTokenRequest

import scala.collection.mutable

object CNFTCommonSteps {

    trait ScenarioContext

    trait WalletsContext extends ScenarioContext {

        val wallets: mutable.Map[String, CNFTWalletSpec[Result]] = mutable.HashMap.empty

        val walletIdentities: mutable.Map[String, String] = mutable.HashMap.empty[String, String]

        def getWalletIdentityByClientName(clientName: String): Result[String] =
            walletIdentities.get(clientName).toRight(s"Unknown client '$clientName'")

        def withWallet(clientName: String)(f: CNFTWalletSpec[Result] => Unit): Unit =
            f(
                wallets
                    .get(clientName)
                    .toRight(s"Unknown client '$clientName'")
                    .orFail(s"Unknown client")
            )

    }

    trait TokenTypesContext extends ScenarioContext {

        val tokenTypeIds: mutable.Map[String, String] = mutable.HashMap.empty

        def withTokenType(tokenTypeName: String)(f: String => Unit): Unit =
            f(
                tokenTypeIds
                    .get(tokenTypeName)
                    .toRight(s"Unknown token type name '$tokenTypeName'")
                    .orFail("Unknown token type")
            )

    }

    trait TokensContext extends ScenarioContext {

        val tokenIds: mutable.Map[String, String] = mutable.HashMap.empty

        def withToken(tokenName: String)(f: String => Unit): Unit =
            f(
                tokenIds
                    .get(tokenName)
                    .toRight(s"Unknown token name '$tokenName'")
                    .orFail("Unknown token")
            )

    }

    def thereIsAClient(
        context: ScenarioContext with WalletsContext,
        clientName: String
    ): Unit = {
        val crypto = createWalletCrypto
        registerMemberHook(crypto)
        val wallet = Chain.newWallet(crypto).orFail(s"Failed to create wallet for '$clientName'")
        val walletIdentity = wallet.getIdentity.orFail(s"Failed to get wallet identity for '$clientName'")
        context.wallets.put(clientName, wallet)
        context.walletIdentities.put(clientName, walletIdentity)
        logger.info(s"Created wallet for client $clientName")
    }

    def clientRegistersTokenType(
        context: ScenarioContext with WalletsContext with TokenTypesContext,
        clientName: String,
        tokenTypeName: String,
        gene: Option[Gene] = None
    ): Unit =
        context.withWallet(clientName) { wallet =>
            (for {
                tokenTypeId <- wallet.createId
                dnaChange =
                    Collection(gene).collect {
                        case Some(gene) => gene
                    }
            } yield {
                context.tokenTypeIds.put(tokenTypeName, tokenTypeId)
                CNFTUtil.registerEmittableTokenType(wallet, tokenTypeId, tokenTypeName, dnaChange).value
            })
                .orFail("Failed to create id for token type")
            ()
        }

    def clientCreatesIdForTokenOfType(
        context: ScenarioContext with WalletsContext with TokenTypesContext with TokensContext,
        clientName: String,
        tokenName: String,
        tokenTypeName: String
    ): Unit =
        context.withWallet(clientName) { wallet =>
            context.withTokenType(tokenTypeName) { tokenTypeId =>
                (for {
                    tokenId <- wallet.createTokenId(tokenTypeId)
                } yield context.tokenIds.put(tokenName, tokenId))
                    .orFail("Failed to create token id")
                ()
            }
        }

    def clientIssuesTokenForClientWithValue(
        context: ScenarioContext with WalletsContext with TokensContext,
        issuerName: String,
        clientName: String,
        tokenName: String,
        tokenValue: String
    ): Unit =
        context.withWallet(issuerName) { issuerWallet =>
            context.withWallet(clientName) { clientWallet =>
                context.withToken(tokenName) { tokenId =>
                    (for {
                        owner <- clientWallet.createSingleOwnerAddress
                        to <- clientWallet.getIdentity
                        body = TokenContent(Collection(tokenValue))
                        request = WalletIssueTokenRequest(tokenId, owner, body, to)
                        _ <- issuerWallet.issue(Collection(request))
                    } yield ())
                        .orFail(s"Failed to issue token '$tokenName' for client '$clientName'")
                    ()
                }
            }
        }

}
